### ISSUE TITLE
Feature req.: possible Bitmap to Outline solution?

### ISSUE THREAD

**Issue by DKqwerty** [fontforge/fontforge#5124]
- [x] No native way to quickly convert bitmap fonts (of any format) to outlines. Reasons for this being difficult are fairly obvious, but broadly-speaking would certainly be an amazingly helpful feature.
- [x] Currently, the only method I can find for FF is [tracing with third-party software](https://fontforge.org/docs/techref/autotrace.html) (and I assume this method is outdated anyways).
- [ ] Rather than trace the entire set of outlines per glyph (which can get really messy with particular glyphs or font styles), would it not be easier to instead literally automate parsing each 1-bit glyph's map as a simple grid and when a pixel is "‍1‍"‍/‍"‍on‍" just place a basic monolithic square in that same position of the same implied grid on an outline glyph? Each "pixel" box would be 1‍/‍w × 1‍/‍h of each glyph, allowing for the blocky, pixelated appearance of the original but as "outlines". (Then at least some of the inevitably redundant/unnecessary stokes could also be removed prior to export with some clever algorithms.)
- [ ] Further, the "pixel" boxes could have a "dot pitch" slider, such that increasing the effective dot pitch shrinks the boxes font‍-‍wide around their centers, creating a tunable "LCD" effect.

Please note, I do understand that this is **not** an *efficient* method relative to the input‍/‍output size: creating a bunch of monolithic boxes per-glyph is creating, eg. 200 points for a fully-black glyph in a 5x10 bitmap font alone (assuming redundant stokes aren't combined‍/‍removed). But since we'd be measuring that difference in KB for older fonts that might not even go beyond U+00FF, I could see it being a potentially-useful tool except when output is to be used on highly space‍-‍constrained systems (like a Pi0 or something).

Just a thought I wanted to share. Obviously I've seen other similar requests for the ability to convert BMP to TTF, but none which suggest this possible solution. But I'm sure I'm still not the first to think of this, so feel free to tell me how dumb and impractical this method would be. Thanks for reading.

**Comment by ctrlcctrlv** [comment]
Please try https://github.com/ctrlcctrlv/bitmapfont2otb/

**Comment by Tynach** [comment]
@ctrlcctrlv, that does not turn a bitmap font into an outline font. It turns a bitmap font from one format into another format.

@DKqwerty, what I do is export the font as .otb, then *import* the font and tell it to import it into the background. I also set the em size to match the pixel size; so if it's a 12px font, I set the font to have an EM size of 12. This makes the font technically invalid (has to be at least 16, for some reason), but it makes tracing the bitmaps much easier, since everything snaps to the bitmap's pixel grid perfectly. I then manually trace the bitmaps.

This is for my own personal font project, though. For existing bitmap fonts with many glyphs, I don't know of any automated solution.

**Comment by ctrlcctrlv** [comment]
> @ctrlcctrlv, that does not turn a bitmap font into an outline font. It turns a bitmap font from one format into another format.

Not true.
![image](https://user-images.githubusercontent.com/838783/197208144-8112e542-6dc4-46aa-bc4b-23b17c136d8f.png)

**Comment by Tynach** [comment]
> Not true.

Huh. While the geometry it generates isn't the greatest, I stand corrected!

Can it turn an .otb file without outlines, into one *with* outlines, while preserving other OpenType tables like kerning, anchors, etc.?

**Comment by DKqwerty** [comment]
Thanks for the replies. I'll give the proposed solution a try when I get a free moment.

_For reference, I'm attempting to exactingly preserve old OS bitmap system fonts—in particular Chicago from the final shipped bitmap version—under modern font containers. The idea here is to create both a true bitmap (for niche use) and then an identical TTF (for generalized resizable usage)._

**Comment by ctrlcctrlv** [comment]
@Tynach It could be tweaked to do so pretty easily I think.

There's nothing wrong with the geometry. CFF subroutinization means it doesn't matter for the most part, and even if a simplified form were smaller, simplification eliminates variable font opportunities. Given how small these fonts are anyway, it's not an issue.

**Comment by ctrlcctrlv** [comment]
@DKqwerty Done properly the OpenType .otb format can solve both use cases.

![image](https://user-images.githubusercontent.com/838783/197261924-f4f46531-322c-4074-8e21-d6c3cb28b2d0.png)

Indeed, even multiple strikes can be handled with OpenType (TrueType) Collections (OTC/TTC), one face index per size.

**Comment by Tynach** [comment]
@ctrlcctrlv, it matters in certain niche uses of the font, such as when trying to make a 3D object with the font 'engraved' on the surface. You'd end up with a bunch of lines intersecting everywhere in that grid pattern.

Also, I thought 'OpenType Bitmap' *only* had bitmaps, but an ordinary OpenType font can have both outline *and* bitmaps? I've been outputting them as separate files for my own project; an .otb for just the bitmaps, and an .otf for both.

**Comment by ctrlcctrlv** [comment]
@Tynach Sorry I still don't think that's a problem. In niche use cases e.g. 3D you can just remove internal geometry. Blender can do this f.e.

**Comment by Tynach** [comment]
@ctrlcctrlv, that's true. What about with font hinting, though? Does it work alright with FreeType's autohinter, for example?

**Comment by ctrlcctrlv** [comment]
I haven't had any problem. If there were one that would be a bug in FreeType.

**Comment by ctrlcctrlv** [comment]
I work on fonts so I know how to check the autohinter's effect on a font. Here's how if you're curious:
```bash
sudo apt install freetype2-demos 
git clone ssh://git@github.com/ctrlcctrlv/clR8x8
cd clR8x8/
ftview 8 clR8x8.otb 
```

at 8px:
![image](https://user-images.githubusercontent.com/838783/200461724-341b7763-843e-4d80-ba34-e342160e58c9.png)

at 16px:
![image](https://user-images.githubusercontent.com/838783/200461835-bffb4ef1-60e9-4b23-9f0e-5c1f46a78567.png)

Now, at erroneous sizes, you get ugly output, e.g. 14px:
![image](https://user-images.githubusercontent.com/838783/200461944-bc53bbb1-cb18-4f43-b9bd-6d1acdc2edb7.png)

or 23px:
![image](https://user-images.githubusercontent.com/838783/200462325-bcb6c60d-4b00-4ce1-a433-bdb88a04fab7.png)

But it's not ugly due to the way I do outlines. It's ugly due to not being a multiple of eight.

**Comment by ctrlcctrlv** [comment]
(But anyway if the hinting becomes a real issue I could just add actual hints so that FreeType won't fall back to the autohinter, I still wouldn't have to change the outlines.)

**Comment by iorsh** [comment]
The requested feature is very specific, and I think it firmly belongs to the plugin domain. A Python script to achieve this would not be particularly complex. While raster glyph data can't be accessed directly, the pseudocode could go like that:

- for each glyph
    - export a glyph to a bitmap file
    - load it with external module, like tkinter.PhotoImage
    - for each pixel draw an outline with a Glyph Pen
